services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthenticationManagerImpl.java

```diff
@@ -18,33 +20,37 @@
 
 	public static final int PASSWORD_MIN_LENGTH = 8;
 
+	public static final long LOCK_TIMOUTE_SEC = 60*60*1000;
+
+	public static final int MAX_CONCURRENT_LOCKS = 10;
+
 	@Autowired
 	private AuthenticationDAO authDAO;
-	
 	@Autowired
 	private UserGroupDAO userGroupDAO;
+	@Autowired
+	private MemoryCountingSemaphore usernameThrottleGate;
 	
 	public AuthenticationManagerImpl() { }
-	
-	/**
-	 * For unit testing
-	 */
-	public AuthenticationManagerImpl(AuthenticationDAO authDAO, UserGroupDAO userGroupDAO) {
-		this.authDAO = authDAO;
-		this.userGroupDAO = userGroupDAO;
-	}
-	
+
 	@Override
 	@WriteTransaction
 	public Session authenticate(long principalId, String password, DomainType domain) throws NotFoundException {
-		// Check the username password combination
-		// This will throw an UnauthorizedException if invalid
-		if (password != null) {
-			byte[] salt = authDAO.getPasswordSalt(principalId);
-			String passHash = PBKDF2Utils.hashPassword(password, salt);
-			authDAO.checkUserCredentials(principalId, passHash);
+		// acquire a lock for throttling password attacks
+		String lockToken = usernameThrottleGate.attemptToAcquireLock(""+principalId, LOCK_TIMOUTE_SEC, MAX_CONCURRENT_LOCKS);
+		if (lockToken != null) {
+			// Check the username password combination
+			// This will throw an UnauthorizedException if invalid
+			if (password != null) {
+				byte[] salt = authDAO.getPasswordSalt(principalId);
+				String passHash = PBKDF2Utils.hashPassword(password, salt);
+				authDAO.checkUserCredentials(principalId, passHash);
+			}
+			usernameThrottleGate.releaseLock(""+principalId, lockToken);
+			return getSessionToken(principalId, domain);
+		} else {
+			throw new RuntimeException(AuthorizationConstants.REASON_TOO_MANY_CONCURRENT_REQUESTS);
```

Comment:
Why throw a RuntimeException?  This error should map to 429 Too Many Requests.  It looks like 429 is currently mapped to TooManyRequestsException.